File: <100/042/042.cpp
#include<iostream>
#include<fstream>
using namespace std;


string line, filename = "042.txt";


bool has_inverse(int n, int (*f)(int)){
  int ok = 1e3, ng = 0;
  while(ok - ng > 1){
    int mid = (ok + ng) / 2;
    if(f(mid) >= n) ok = mid;
    else            ng = mid;
  }

  return f(ok) == n; 
}


int triangle_prod(int n){ return n * (n + 1) / 2; }
bool is_triangle(int n){ return has_inverse(n, triangle_prod); }


int name_to_num(string s){
  int result = 0;
  for(auto c : s)if(isupper(c)){
    result += c - 'A' + 1;
  }

  return result;
}


int main(){
  int answer = 0;
  
  ifstream file(filename);
  while(getline(file, line, ',')){
    int n = name_to_num(line);
    if(is_triangle(n)){
      ++answer;
    }
  }

  cout << answer << endl;
  return 0;
}

File: <100/042/042.hs
import System.IO ( hGetContents, openFile, IOMode(ReadMode), hClose )
import Text.Regex ( mkRegex, splitRegex )
import Data.Char ( isAsciiUpper, ord )


main :: IO ()
main = do
  handle   <- openFile filename ReadMode
  contents <- hGetContents handle
  print . length . filter isTriangle . map (convert . extractCapitals) . splitComma $ contents
  hClose handle


-- make "SKY" to 55 = 19 + 11 + 25
convert :: String -> Int
convert [] = 0
convert (c:cs) = ord' + convert cs
    where ord' = ord c - ord 'A' + 1


isTriangle :: Integral int => int -> Bool
isTriangle = hasInverse triangleProd


-- returns if n can be produced by monotically increasing function f :: int -> int
hasInverse :: Integral int => (int -> int) -> int -> Bool
hasInverse f n = n == f lb
    where lb = lowerBoundInverse' 0 (10^9) f n


-- returns Sup{x | f x <= n, x in n}
lowerBoundInverse' :: Integral int => int -> int -> (int -> int) -> int -> int
lowerBoundInverse' smaller bigger f n
    | gap == 1   = smaller
    | f mid <= n = lowerBoundInverse' mid bigger f n
    | otherwise  = lowerBoundInverse' smaller mid f n
    where
        gap = bigger - smaller
        mid = (bigger + smaller) `div` 2


triangleProd :: Integral int => int -> int
triangleProd n = n*(n + 1) `div` 2


----------------------- supplemental -----------------------
filename :: String
filename   = "042.txt"


splitComma :: String -> [String]
splitComma = splitRegex (mkRegex ",")


extractCapitals :: String -> String
extractCapitals [] = []
extractCapitals (c:cs)
    | isCapital = c : extractCapitals cs
    | otherwise = extractCapitals cs
    where isCapital = isAsciiUpper c
